src/lib/logstore/log_dev.cpp

```diff
@@ -580,7 +583,242 @@ void LogDev::rollback(logstore_id_t store_id, logid_range_t id_range) {
+    for (auto it{std::begin(m_unopened_store_io)}; it != std::end(m_unopened_store_io); ++it) {
+        LOGINFO("skip log entries for store id {}-{}, ios {}", m_logdev_id, it->first, it->second);
+    }
+    m_unopened_store_io.clear();
+
+    // If there are any unopened storeids found, loop and check again if they are indeed open later. Unopened log store
+    // could be possible if the ids are deleted, but it is delayed to remove from store id reserver. In that case,
+    // do the remove from store id reserver now.
+    // TODO: At present we are assuming all unopened store ids could be removed. In future have a callback to this
+    // start routine, which takes the list of unopened store ids and can return a new set, which can be removed.
+    {
+        folly::SharedMutexWritePriority::WriteHolder holder(m_store_map_mtx);
+        for (auto it{std::begin(m_unopened_store_id)}; it != std::end(m_unopened_store_id);) {
+            if (m_id_logstore_map.find(*it) == m_id_logstore_map.end()) {
+                // Not opened even on second time check, simply unreserve id
+                unreserve_store_id(*it);
+            }
+            it = m_unopened_store_id.erase(it);
+        }
+
+        // Also call the logstore to inform that start/replay is completed.
+        if (!format) {
+            for (auto& p : m_id_logstore_map) {
+                auto& lstore{p.second.m_log_store};
+                if (lstore && lstore->get_log_replay_done_cb()) {
+                    lstore->get_log_replay_done_cb()(lstore, lstore->seq_num() - 1);
+                    lstore->truncate(lstore->truncated_upto());
+                }
+            }
+        }
+    }
+}
+
+std::shared_ptr< HomeLogStore > LogDev::create_new_log_store(bool append_mode) {
+    auto const store_id = reserve_store_id();
+    std::shared_ptr< HomeLogStore > lstore;
+    lstore = std::make_shared< HomeLogStore >(shared_from_this(), store_id, append_mode, 0);
+
+    {
+        folly::SharedMutexWritePriority::WriteHolder holder(m_store_map_mtx);
+        const auto it = m_id_logstore_map.find(store_id);
+        HS_REL_ASSERT((it == m_id_logstore_map.end()), "store_id {}-{} already exists", m_logdev_id, store_id);
+        m_id_logstore_map.insert(std::make_pair<>(store_id, logstore_info_t{lstore, nullptr, append_mode}));
+    }
+    LOGINFO("Created log store id {}-{}", m_logdev_id, store_id);
+    return lstore;
+}
+
+void LogDev::open_log_store(logstore_id_t store_id, bool append_mode, const log_store_opened_cb_t& on_open_cb) {
+    folly::SharedMutexWritePriority::WriteHolder holder(m_store_map_mtx);
+    const auto it = m_id_logstore_map.find(store_id);
+    HS_REL_ASSERT((it == m_id_logstore_map.end()), "store_id {}-{} already exists", m_logdev_id, store_id);
+    LOGINFO("Opening log store id {}-{}", m_logdev_id, store_id);
+    m_id_logstore_map.insert(std::make_pair<>(store_id, logstore_info_t{nullptr, on_open_cb, append_mode}));
+}
```

Comment:
I think depending on which PR goes first, there will some clash with my PR on Restartability of ReplDev and this one, since we made the open_log_store to return a future.

Reply:
Yes rebased to latest.